flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java

```diff
@@ -116,52 +115,58 @@ protected Throwable getFailureCause() {
 
 	public abstract boolean isReleased();
 
-	/**
-	 * Gets the number of non-event buffers in this subpartition.
-	 *
-	 * <p><strong>Beware:</strong> This method should only be used in tests in non-concurrent access
-	 * scenarios since it does not make any concurrency guarantees.
-	 */
-	@VisibleForTesting
-	public int getBuffersInBacklog() {
-		return buffersInBacklog;
-	}
-
 	/**
 	 * Makes a best effort to get the current size of the queue.
 	 * This method must not acquire locks or interfere with the task and network threads in
 	 * any way.
 	 */
 	public abstract int unsynchronizedGetNumberOfQueuedBuffers();
 
+	/**
+	 * Gets the number of non-event buffers in this subpartition.
+	 */
+	public abstract int getBuffersInBacklog();
+
+	/**
+	 * @param lastBufferAvailable whether the last buffer in this subpartition is available for consumption
+	 * @return the number of non-event buffers in this subpartition
+	 */
+	protected int getBuffersInBacklog(boolean lastBufferAvailable) {
+		assert Thread.holdsLock(buffers);
+
+		if (lastBufferAvailable) {
+			return buffersInBacklog;
+		} else {
+			return Math.max(buffersInBacklog - 1, 0);
+		}
+	}
+
 	/**
 	 * Decreases the number of non-event buffers by one after fetching a non-event
 	 * buffer from this subpartition (for access by the subpartition views).
-	 *
-	 * @return backlog after the operation
 	 */
-	public int decreaseBuffersInBacklog(Buffer buffer) {
+	public void decreaseBuffersInBacklogUnsafe(boolean isBuffer) {
```

Comment:
I misunderstood the semantic of `Unsafe` before. I define the `Unsafe` based on the external calling not internal implementation of `decreaseBuffersInBacklog`. In other words, if external method calls `decreaseBuffersInBacklog`, it should confirm the buffers are synchronized in the external method to call this safe, otherwise external method should call `decreaseBuffersInBacklogUnsafe`.

Now I understand your point of definition, and it seems more reasonable based on internal confirmation of `decreaseBuffersInBacklog`. I would revert this change as before.

Reply:
I misunderstood the semantic of `Unsafe` before. I define the `Unsafe` based on the external calling not internal implementation of `decreaseBuffersInBacklog`. In other words, if external method calls `decreaseBuffersInBacklog`, it should confirm the buffers are synchronized in the external method to call this safe, otherwise external method should call `decreaseBuffersInBacklogUnsafe`.

Now I understand your point of definition, and it seems more reasonable based on internal confirmation of `decreaseBuffersInBacklog`. I would revert this change as before.